include/rnp/rnp2.h

```diff
@@ -0,0 +1,283 @@
+
+rnp_result_t rnp_keyring_save_to_mem(rnp_keyring_t ring, uint8_t *buf[], size_t *buf_len);
+rnp_result_t rnp_keyring_free(rnp_keyring_t *ring);
+
+/* TODO: keyring iteration */
+
+/** generate a key (or pair of keys)
+ *
+ * @param primarykey the primary key pointer that will be populated,
+ *        must not be NULL
+ * @param subkey the subkey key pointer that will be populated,
+ *        must not be NULL
+ * @param jsondata the JSON string that describes the key generation,
+ *        must not be NULL
+ * @return 0 on success, or any other value on error
+ */
+rnp_result_t rnp_generate_key_json(rnp_key_t * primarykey,
+                                   rnp_key_t * subkey,
+                                   const char *jsondata);
+
+/* Key operations */
+
+/**
+* Export a public key from the keyring
+*/
+#define RNP_EXPORT_FLAG_ARMORED (1U << 0)
+rnp_result_t rnp_export_public_key(rnp_key_t key,
+                                   uint32_t  flags,
+                                   char **   output,
+                                   size_t *  output_len);
+
+/* TODO: export encrypted secret keys */
+
+rnp_result_t rnp_key_get_primary_uid(rnp_key_t key, char **uid);
+rnp_result_t rnp_key_get_uid_count(rnp_key_t key, size_t *count);
+rnp_result_t rnp_key_get_uid_at(rnp_key_t key, size_t idx, char **uid);
+
+/* TODO: should these be NULL-terminated hex strings, or just
+ the raw data? Document which. */
+rnp_result_t rnp_key_get_fprint(rnp_key_t key, char **fprint);
+rnp_result_t rnp_key_get_keyid(rnp_key_t key, char **keyid);
+rnp_result_t rnp_key_get_grip(rnp_key_t key, char **grip);
+
+rnp_result_t rnp_key_is_locked(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_unlock(rnp_key_t key, rnp_passphrase_cb cb, void *app_ctx);
+
+rnp_result_t rnp_key_is_protected(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_protect(rnp_key_t key, rnp_passphrase_cb cb, void *app_ctx);
+rnp_result_t rnp_key_unprotect(rnp_key_t key, rnp_passphrase_cb cb, void *app_ctx);
+
+rnp_result_t rnp_key_is_primary_key(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_is_subkey(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_is_secret(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_is_public(rnp_key_t key, bool *result);
+
+/* TODO: function to add a userid to a key */
+
+/* Signature/verification operations */
+
+/**
+* Generate an embedded signature (the output will include the
+* message contents)
+* @param keyring the keyring
+* @param ident the key to sign with
+* @param hash_fn the hash function to use
+* @param msg the message to sign
+* @param msg_len the length of msg in bytes
+* @param sig on success, the output signature buffer
+* @param sig_len on success, the length of sig in bytes
+*/
+rnp_result_t rnp_sign(rnp_keyring_t keyring,
+                      const char *  ident,
+                      const char *  hash_fn,
+                      bool          clearsign,
```

Comment:
Imho it would be logical to separate cleartext signing from 'normal' signing since they are a bit different things. Also such interface will not allow signing with multiple keys.

Reply:
> Imho it would be logical to separate cleartext signing from 'normal' signing since they are a bit different things.

Maybe, I'll take a look.

> Also such interface will not allow signing with multiple keys.

We don't really support this internally at the moment though. In the future we could add some kind of "sign operation" concept and "add signer" "add data" operations, etc. We're not aiming to get everything done perfectly on the first round here.